### PR TITLE
Added a Learning Rate Schedule for the BERT Example

### LINKED ISSUE #84: Add a proper learning rate schedule to the BERT example
We should add learning rate schedule matching the [original BERT implementation](https://github.com/google-research/bert/blob/master/optimization.py#L32) to our BERT example in `/examples/bert`.

### PR DESCRIPTION
**Pull request by Stealth-py**
Fixes #84
This PR is ready for review!

### CODE REVIEW

File: examples/bert/run_pretraining.py [hunk 1]
```diff
@@ -291,6 +301,53 @@ def train_step(self, data):
...
+            num_warmup_steps,
+            num_train_steps
+        ):
+        self.threshold_learning_rate = learning_rate
+        self.warmup_steps = num_warmup_steps
+        self.train_steps = num_train_steps
+        
+    
+    def __call__(self, step):
+        """
```
Comment (mattdangerw) [inline examples/bert/run_pretraining.py:321]: I think we can probably just have one docstring for the class (and non for call).

File: examples/bert/run_pretraining.py [hunk 2]
```diff
@@ -291,6 +301,53 @@ def train_step(self, data):
         self.nsp_accuracy.update_state(data["next_sentence_labels"], nsp_preds)
         return {m.name: m.result() for m in self.metrics}
 
+class AdamLRSchedule(keras.optimizers.schedules.LearningRateSchedule):
```
Comment (mattdangerw) [inline examples/bert/run_pretraining.py:304]: rename LinearDecayWithWarmup

File: examples/bert/run_pretraining.py [hunk 3]
```diff
@@ -291,6 +301,53 @@ def train_step(self, data):
...
+    """
+    Implements a Learning Rate Scheduler for Adam with Linear Warmup and
```
Comment (mattdangerw) [inline examples/bert/run_pretraining.py:306]: In general dostring should have a short oneliner:

Implements a learning rate schedule for BERT.

Optionally followed by more text after an empty line.

File: examples/bert/run_pretraining.py [hunk 4]
```diff
@@ -291,6 +301,53 @@ def train_step(self, data):
...
+        """
+        final_lr = tf.cast(self.threshold_learning_rate, dtype = tf.float32)
+        warmup = tf.cast(self.warmup_steps, dtype = tf.float32)
+        training = tf.cast(self.train_steps, dtype = tf.float32)
+        
+        is_warmup = step < warmup
+
+        #Linear Decay will be implemented if current step is not less than
+        #`num_warmup_steps`
+        learning_rate = tf.keras.optimizers.schedules.PolynomialDecay(
```
Comment (mattdangerw) [inline examples/bert/run_pretraining.py:335]: I think this actually distracts from what's going on here. Can't we do something like this (though also converting to floats)?

```
if step < warmup_steps:
    return step / warmup_steps
else:
    return max(0.0, (train_steps - step) / (train_steps - warmup_steps))
```

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks! Left some comments

File: examples/bert/run_pretraining.py [hunk 5]
```diff
@@ -348,19 +405,28 @@ def main(_):
...
     # Wrap with pretraining heads and call fit.
     pretraining_model = BertPretrainer(model)
     pretraining_model.compile(
         # TODO(mattdangerw): Add AdamW and a learning rate schedule.
-        optimizer=keras.optimizers.Adam(learning_rate=FLAGS.learning_rate)
+        optimizer=keras.optimizers.Adam(learning_rate=learning_rate_schedule)
     )
     # TODO(mattdangerw): Add TPU strategy support.
-    pretraining_model.fit(dataset, epochs=FLAGS.epochs)
+    steps_per_epoch = FLAGS.num_train_steps/FLAGS.epochs
```
[the remark this answers is not in the document]
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:421]: Yeah, I messed up a bit there. It should be floor division for `step_per_epoch`, yes. I'll make the changes in a bit.

File: examples/bert/run_pretraining.py [hunk 6]
```diff
@@ -61,6 +61,16 @@
...
 
+flags.DEFINE_integer(
+    "num_warmup_steps",
+    1e4,
+    "The number of warmup steps during which the learning rate will increase "
+    "till a threshold.")
+
+flags.DEFINE_integer(
+    "num_train_steps",
+    3e4,
```
[the remark this answers is not in the document]
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:72]: Yeah, makes sense. I'll make the changes necessary.

File: examples/bert/run_pretraining.py (hunk 1 above)
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:321]: Alright, would you like me to use a single line comment for the `call` method or is it fine if it has none?

File: examples/bert/run_pretraining.py (hunk 4 above)
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:335]: That is a better and a shorter way to do this, yeah. I was using this in the beginning but then I wondered if using `PolynomialDecay` would be the same or not, and just ended up using this. I'll change it and resolve this in a bit.

File: .gitignore [hunk 7]
```diff
@@ -15,3 +15,5 @@ dist/
 
 # Windows-specific build files
 build/
+
+venv/
```
[the remark this answers is not in the document]
Comment (Stealth-py) [reply .gitignore:19]: Alright, I'll add it

File: examples/bert/run_pretraining.py [hunk 8]
```diff
@@ -291,6 +301,53 @@ def train_step(self, data):
...
+    def __init__(self,
+            learning_rate,
+            num_warmup_steps,
+            num_train_steps
+        ):
+        self.threshold_learning_rate = learning_rate
+        self.warmup_steps = num_warmup_steps
+        self.train_steps = num_train_steps
+        
+
```
[the remark this answers is not in the document]
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:319]: Alright

File: examples/bert/run_pretraining.py (hunk 2 above)
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:304]: Yeah, it's better to name it this way. Would also help reduce the extra comments in the docstring.

File: examples/bert/run_pretraining.py (hunk 3 above)
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:306]: Alright, I'll change it

**Comment by Stealth-py** [comment]
Thanks for the review @mattdangerw!
I've made the necessary changes and resolved the comments above.

File: examples/bert/run_pretraining.py [hunk 9]
```diff
@@ -348,19 +390,28 @@ def main(_):
...
+    #Implemented Learning Rate Schedule
+    learning_rate_schedule = LinearDecayWithWarmup(
+        learning_rate = FLAGS.learning_rate,
+        num_warmup_steps = FLAGS.num_warmup_steps,
+        num_train_steps = FLAGS.num_train_steps)
+    
     # Wrap with pretraining heads and call fit.
     pretraining_model = BertPretrainer(model)
     pretraining_model.compile(
         # TODO(mattdangerw): Add AdamW and a learning rate schedule.
```
Comment (mattdangerw) [inline examples/bert/run_pretraining.py:402]: You could remove the learning rate schedule bit from this TODO.

File: examples/bert/run_pretraining.py [hunk 10]
```diff
@@ -291,6 +301,38 @@ def train_step(self, data):
...
+    and for the steps post that, till the `num_train_steps` it will implement 
+    a linear decay.
+    """
+    
+    def __init__(self,
+            learning_rate,
+            num_warmup_steps,
+            num_train_steps
+        ):
+        self.threshold_learning_rate = learning_rate
```
Comment (mattdangerw) [inline examples/bert/run_pretraining.py:318]: Keep the same name as argument...

self.learning_rate = learning_rate

File: examples/bert/run_pretraining.py [hunk 11]
```diff
@@ -291,6 +301,38 @@ def train_step(self, data):
...
+        self.warmup_steps = num_warmup_steps
+        self.train_steps = num_train_steps
+    
+    def __call__(self, step):
+        final_lr = tf.cast(self.threshold_learning_rate, dtype = tf.float32)
```
Comment (mattdangerw) [inline examples/bert/run_pretraining.py:323]: Maybe call this peak_lr? final_lr is actually 0.0.

File: examples/bert/run_pretraining.py [hunk 12]
```diff
@@ -291,6 +301,38 @@ def train_step(self, data):
...
+        warmup = tf.cast(self.warmup_steps, dtype = tf.float32)
+        training = tf.cast(self.train_steps, dtype = tf.float32)
+        
+        is_warmup = step < warmup
+        
+        #Linear Warmup will be implemented if current step is less than
```
Comment (mattdangerw) [inline examples/bert/run_pretraining.py:329]: space after `#` here and next line

**Review by mattdangerw — APPROVED** [review]
Looks good, just a few more style nits, then can merge!

File: examples/bert/run_pretraining.py (hunk 9 above)
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:402]: Ah, yes. Alright!

File: examples/bert/run_pretraining.py (hunk 10 above)
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:318]: Okay, I'll change it

File: examples/bert/run_pretraining.py (hunk 11 above)
Comment (Stealth-py) [reply examples/bert/run_pretraining.py:323]: Yeah, makes sense. I was also thinking of changing it to threshold itself.

**Comment by Stealth-py** [comment]
Thank you! I've resolved the issues and pushed :D

**Comment by mattdangerw** [comment]
@Stealth-py thanks! Can you fix formatting errors? https://github.com/keras-team/keras-nlp/runs/5979856948?check_suite_focus=true

**Comment by Stealth-py** [comment]
Yeah, I'm sorry I should've ran it before pushing into the main directory.
It should be fixed now I think.

**Comment by mattdangerw** [comment]
@Stealth-py thanks! I will merge now with a few last copy edits.

**Comment by Stealth-py** [comment]
Alright!
Should I close this PR?

**Comment by mattdangerw** [comment]
@Stealth-py no need! We generally merge through github which will close. Thanks!
